### PR TITLE
os/kernel/task: Fix for multiple exit scenario via task_exithook()

### PR DESCRIPTION
**Pull request by rish-sg**
Description:
This patch prevents multiple exit of task via task_exithook(), in SMP case.

----------------------------------------------------------------------

AS-IS:
TCB_FLAG_EXIT_PROCESSING is bit set in tcb->flag to prevent multiple exit. But the checker and setter code are not protected and are being called at different places.
Checker: at the start of the function to return early if the flag is set. Setter: flag is set at the end of task_exithook(). This difference in exection of checker and setter opens a window of race where one process sets it, but another process has already passed the checker point.
In this race, both the processes may attempt to kill the same task, which leads to a multiple-exit scenario.

----------------------------------------------------------------------

LOGS:
log_csi_config: enable: 0
wifi_csi_mq_close: MQ cltask_sigchild: task: csifw_task_Main, tg_nchildren = 1 task_sigchild: task: task_manager, tg_nchildren = 0

security level: 0
=========================================================== Assertion details
=========================================================== Assertion failed CPU0 at file: task/task_exithook.c line 339 task: task_manager pid: 17 print_assert_detail: Assert location (PC) : 0x0e024eff check_assert_location: Code asserted in normal thread! =========================================================== Asserted task's stack details
=========================================================== check_sp_corruption: Current SP is User Thread SP: 60154490

----------------------------------------------------------------------

Fix: We move the setter code to the beginning of task_exithook() (just after the checker) and protect them within critical section to ensure atomicity.

----------------------------------------------------------------------

### CODE REVIEW

File: os/kernel/task/task_exithook.c [hunk 1]
```diff
@@ -553,10 +553,31 @@ void task_exithook(FAR struct tcb_s *tcb, int status, bool nonblocking)
 	 * that bit is set, then just exit doing nothing more..
 	 */
 
+#ifdef CONFIG_SMP
```
Comment (sunghan-chang) [inline os/kernel/task/task_exithook.c:556]: The SMP conditional is necessary? What about context switching here in non smp?

**Comment by seokhun-eom24** [comment]
> Automated nightly Codex PR review
> PR: #7288 — os/kernel/task: Fix for multiple exit scenario via task_exithook()
> PR URL: https://github.com/Samsung/TizenRT/pull/7288
> GitHub updatedAt: 2026-05-13T12:47:16Z
> Refreshed (UTC): 2026-05-13 17:47:20 UTC
> Refreshed (KST): 2026-05-14 02:47:20 KST
> Comment model: single evolving Codex-authored PR comment
# PR #7288 - os/kernel/task: Fix for multiple exit scenario via task_exithook()

> **This review done by codex. AI reviews can be inaccurate.**  
> **UTC:** 2026-05-13 17:46 UTC  
> **KST (UTC+9):** 2026-05-14 02:46 KST

---

**Repository**: `Samsung/TizenRT`

**Base -> Head**: `master` -> `pr/7288`

**HEAD Commit**: `f3ada178ea209d9ed6a4ad0d68de45f5cda520c3`

**Scope**: Static review of task-exit race handling in `os/kernel/task/task_exithook.c` and `os/kernel/task/task_terminate_unloaded.c`.

---

## Review Summary

| Category | Status |
| --- | --- |
| Build / Compile Status | Not run |
| Functional Correctness | Request changes |
| Validation Coverage | Not provided |

**Final Verdict: Request Changes**

---

## Must-Fix Issues

<details>
<summary><strong>1. Duplicate-exit guard still leaves the original race window open</strong></summary>

<br>

| Item | Details |
| --- | --- |
| **Location** | `os/kernel/task/task_exithook.c:558` |
| **Severity** | High |
| **Type** | Functional correctness / SMP race |

**Problem**  
The PR protects the `TCB_FLAG_EXIT_PROCESSING` check with `enter_critical_section()`, but it releases the critical section before setting the flag. The flag is not set until line 638, after debug termination capture, cancellation flag updates, optional `atexit`/`on_exit` callbacks, and `task_recover()`. A second CPU can therefore enter `task_exithook()` after the first CPU leaves line 566 and before the first CPU reaches line 638, observe the flag still clear, and run the same exit cleanup path.

**Impact**
- The PR does not actually close the multiple-exit race described in its own PR body.
- Two callers can still reach cleanup paths such as `task_recover()`, `task_signalparent()`, `task_exitwakeup()`, `binary_manager_remove_binlist()`, `group_leave()`, and `sig_cleanup()` for the same TCB.
- This can preserve the reported assertion/crash class rather than fixing it, especially on SMP where the PR says the race occurs.

**Evidence**
- `os/kernel/task/task_exithook.c:558` - enters a critical section for the guard.
- `os/kernel/task/task_exithook.c:560` - returns only if `TCB_FLAG_EXIT_PROCESSING` is already set.
- `os/kernel/task/task_exithook.c:566` - leaves the critical section without setting `TCB_FLAG_EXIT_PROCESSING`.
- `os/kernel/task/task_exithook.c:638` - sets `TCB_FLAG_EXIT_PROCESSING` much later, after several cleanup operations have already executed.
- `os/kernel/task/task_terminate.c:236` and `os/kernel/task/task_exit.c:263` - `task_exithook()` can be reached through external termination and current-task exit paths, which is the cross-path re-entry case the PR is trying to serialize.

**Required Fix**
Files to edit:
- `os/kernel/task/task_exithook.c` - `task_exithook()`

Change outline:
- Make the guard a single atomic test-and-set operation.
- Set `TCB_FLAG_EXIT_PROCESSING` while still inside the same critical section that checked it.
- Remove the later duplicate assignment at line 638.
- Keep the longer cleanup critical section policy separate from the guard; do not rely on a later assignment to serialize entry.

Example patch:
```diff
diff --git a/os/kernel/task/task_exithook.c b/os/kernel/task/task_exithook.c
@@
 	flags = enter_critical_section();
 
 	if ((tcb->flags & TCB_FLAG_EXIT_PROCESSING) != 0) {
- 
 		leave_critical_section(flags);
 		return;
 	}
 
+	tcb->flags |= TCB_FLAG_EXIT_PROCESSING;
+
 	leave_critical_section(flags);
@@
-	tcb->flags |= TCB_FLAG_EXIT_PROCESSING;
-
 	/* Send the SIGCHILD signal to the parent task group */
```

Inference:
- The exact placement after the guard is inferred from the PR's stated design goal and from the existing `TCB_FLAG_EXIT_PROCESSING` contract. Static review did not run an SMP stress reproducer.

**Validation Method**
- Add or run an SMP stress scenario where one path is already in `task_exithook()` while another path attempts to terminate the same TCB through `task_delete()` / `task_exit()`.
- Instrument `task_exithook()` temporarily and confirm only one caller per TCB passes the guard; later callers must return before `task_recover()`.
- Re-run the original crash scenario from the PR body and confirm the assertion in `task/task_exithook.c` no longer reproduces.

</details>

---

## Nice-to-Have Improvements

<details>
<summary><strong>1. Add explicit validation coverage for the SMP re-entry contract</strong></summary>

<br>

| Item | Details |
| --- | --- |
| **Location** | `os/kernel/task/task_exithook.c:549` |
| **Severity** | Medium |
| **Type** | Validation |

**Problem**  
This patch changes low-level task-exit serialization, but the PR does not include a test, stress command, or documented validation method that proves duplicate entry is blocked at the start of `task_exithook()`.

**Impact**
- A visually small guard change can appear correct while still leaving a race window, as this revision does.
- Future edits to task exit, task deletion, or binary unload cleanup have no regression signal for the duplicate-exit invariant.

**Recommended Action**
- Add a focused SMP regression scenario or a documented stress recipe that forces concurrent termination of the same task and records that only one caller executes the cleanup body.
- At minimum, include the original reproduction command/log source and the fixed-run result in the PR validation notes.

**Example Validation Matrix**

| Area | Example Target / Check | Result |
| --- | --- | --- |
| SMP task exit | Concurrent `task_delete()` / exiting-task path for same TCB | Not run in this review |
| Regression crash | Original `task_manager` assertion scenario | Not run in this review |
| Static review | Guard and flag assignment ordering | Failed: race remains |

</details>

---

<details>
<summary><strong>Notable Improvements</strong></summary>

<br>

### OK The patch targets the right TCB-level serialization point

- `TCB_FLAG_EXIT_PROCESSING` is the existing per-TCB guard used by `task_exithook()`, so fixing the race at this flag is the right subsystem boundary.
- The PR also recognizes that the issue is SMP-sensitive and that the check/update must be protected from concurrent callers.

</details>

---

## Final Assessment

### Must-Fix Summary

- The duplicate-exit check and flag update are still not atomic. The flag must be set inside the same critical section as the check, before any cleanup work runs.

### Nice-to-Have Summary

- Add explicit SMP re-entry validation or at least document the original reproduction and fixed-run evidence.

### Residual Risk / Test Gap

- This was a static review only. No board build, SMP runtime stress test, or crash reproduction was executed in this unattended run.

---

## Final Verdict

> Request Changes

The PR is not merge-ready because the intended race fix still leaves a window where another CPU can pass the guard before `TCB_FLAG_EXIT_PROCESSING` is set. Move the flag assignment into the initial guarded region and validate the duplicate-exit scenario.

File: os/kernel/task/task_exithook.c (hunk 1 above)
Comment (rish-sg) [reply os/kernel/task/task_exithook.c]: Got you. Need to make sure pre-emption is disabled in non-SMP case too. Condition removed.

**Comment by rish-sg** [comment]
Yes, we have checked this, there should be a single task termination handler. 

Similar migration was done on Nuttx and we have found the commit:
Related Nuttx commit: https://github.com/apache/nuttx/commit/eae57cb0e6c2677dcf42e95959a8dd9946889602
But this seems to be part of bigger code changes and simple patch may not be enough.

We need to evaluate all the related code changes and then take a decision which will take time to validate.
Currently only going with this change.

**Comment by seokhun-eom24** [comment]
> Yes, we have checked this, there should be a single task termination handler.
> 
> Similar migration was done on Nuttx and we have found the commit: Related Nuttx commit: [apache/nuttx@eae57cb](https://github.com/apache/nuttx/commit/eae57cb0e6c2677dcf42e95959a8dd9946889602) But this seems to be part of bigger code changes and simple patch may not be enough.
> 
> We need to evaluate all the related code changes and then take a decision which will take time to validate. Currently only going with this change.

Thank you for checking.
Please review and correct this part too in new PR soon.

I think we need to check task_terminate, task_exithook, task_delete, pthread_exit, and exit.

**Review by seokhun-eom24 — APPROVED** [review]

File: os/kernel/task/task_exithook.c [hunk 2]
```diff
@@ -553,10 +553,25 @@ void task_exithook(FAR struct tcb_s *tcb, int status, bool nonblocking)
...
+		leave_critical_section(flags);
 		return;
 	}
 
+	/* This function can be re-entered in certain cases.  Set a flag
+	 * bit in the TCB to note that we have already completed this exit
+	 * processing.
+	 */
+
+	tcb->flags |= TCB_FLAG_EXIT_PROCESSING;
```
Comment (ewoodev) [inline os/kernel/task/task_exithook.c:571]: if Setting it here, We must ensure that the exit hook sequent is performed.

but I think we should consider the case below.
1. the thread is entering terminate itself and pass the line.
2. the context switching is occured.
3. other thread is call task_delete or pthread_cancel.
  -> TCB_FLAG_EXIT_PROCESSING is setted, skip the task_exithook.
  -> pull it out of readytorun queue. 
  => the task_exithook sequence will not perfored.

Comment (rish-sg) [reply os/kernel/task/task_exithook.c]: I agree with your point. For now, we have moved the setting of TCB_FLAG_EXIT_PROCESSING to the lower critical_section() of task_exithook(), this will ensure that the exithook sequence is performed at least once.

But it just reduces the race window and not completely closes it. We have tested it for the issue observed, currently it is working fine.

**Comment by eunwoo-nam** [comment]
> Yes, we have checked this, there should be a single task termination handler.
> 
> Similar migration was done on Nuttx and we have found the commit: Related Nuttx commit: [apache/nuttx@eae57cb](https://github.com/apache/nuttx/commit/eae57cb0e6c2677dcf42e95959a8dd9946889602) But this seems to be part of bigger code changes and simple patch may not be enough.
> 
> We need to evaluate all the related code changes and then take a decision which will take time to validate. Currently only going with this change.

The nuttx changes is including task_recover into enter_critical_section. Could you please check we also need to change?

**Comment by rish-sg** [comment]
The nuttx changes include the complete task_exithook() function inside the critical_section() (ensured by its caller) and it is a part of a bigger change. It will require a complete re-work. Currently, we plan to keep the change as is.

File: os/kernel/task/task_exithook.c
Comment (eunwoo-nam) [inline os/kernel/task/task_exithook.c]: The task_exithook is called from exit, task_terminate.
exit: nonbolocking = false
task_terminate(pthread_cancel): nonblocking = true.

exit case can make context swtiching.
in this timming, if task_terminate is called, the group_leave and sig_cleanup will not be called.

we need to avoid context swtiching in task_exithook using enter_critical_section and sched_lock
and moving flushing api.

please refer
https://github.com/apache/nuttx/commit/3be81e649f5940f4bfd609462354a22c00e48d2c
https://github.com/apache/nuttx/commit/eae57cb0e6c2677dcf42e95959a8dd9946889602

**Review by ewoodev — APPROVED** [review]
